### ISSUE TITLE
No completion suggestions where there should be some 

### ISSUE THREAD

**Issue by claudio-silva** [TypeStrong/atom-typescript#82]
I've been searching for a package that allows me to write typescript with a decent amount of code analysis and autocompletion support. Your's is the best I've found so far. Nice work!
Nevertheless, there are some annoyances that keep me from diving into Atom as my main editor, using your package. They may seem to be small details, but those details do make a difference.
For me, autocompletion is all about API exploration and instant documentation, and not just about completing words. That's why simple text editors (like Sublime Text) are not enough for me; the autocompletion is so basic it's not really useful for serious work.

Now, regarding this issue:
Currently, there are many situations where I get no completion suggestions or I only get them when I type, at least, one character for a identifier (property name, etc).
For example, if I type:
```typescript
function w(x:{test:string}) { return ''; }
w({});
```
I would like to be able to get autocompletion when the cursor is inside the curly brackets, so that I get a list of all available properties valid for the specified object. Currently, I only get a popup when I type the letter `t`, `ctrl+shift+space` does nothing.

Or, after typing a dot, I would like to know all available properties or methods at that location.
On the example above, if I type `w.` on a new line, I get a list of methods for the `Function` type, but if I type `w({test:'a'}).`, I get nothing.

More examples:
```typescript
class A<T>
{
  id (): T
  {
    return null;
  }
}

class B<T> extends A<T>
{
  name (): T
  {
    return null;
  }
}

interface IB extends B<IB>
{}

function x (): IB
{
  return null;
}

x ().
  name ().
  id ().
  name ().
  id ();
```
When typing a dot after `x()` it would be very useful to get both `name` and `id` as suggestions. `ctrl+space` also does nothing.
After each method invocation on the chain, one should be able to get autocompletion. I get none.

One last example:
```typescript
interface AA<T> {
    name?:T;
}
interface BB<T> extends AA<T> {
    age?: number;
}
function y(a: BB<string>) {}


y ( {name: 's', age:3} );
y ( {} )
```
I should be able to get both `name` and `age` as autocompletion suggestions when the cursor is placed inside the curly braces of `y ( {} )`, but I can only get a suggestion if I type `n` or `a`.

**Comment by basarat** [comment]
> I would like to be able to get autocompletion when the cursor is inside the curly brackets, so that I get a list of all available properties valid for the specified object. 

* [ ] I'll raise a bug with the TypeScript team. The language service is returning on results. 
Workaround: insert a newline:

![image](https://cloud.githubusercontent.com/assets/874898/6205335/a3435c4e-b5be-11e4-8cfc-32a36ca6373e.png)

> On the example above, if I type w. on a new line, I get a list of methods for the Function type, but if I type w({test:'a'})., I get nothing

* [ ] Another bug. Thanks for the report. 

> I should be able to get both name and age as autocompletion suggestions when the cursor is placed inside the curly braces of y ( {} ), but I can only get a suggestion if I type n or a

Similar to first bug.

**Comment by claudio-silva** [comment]
@basarat, thank you for your workaround, it works for function arguments, and that's great!
Unfortunately it doesn't work for the `x()` function, i.e. for method chaining.
Also, thanks a lot for taking the time to report this to the TypeScript team!
This is an important issue to me, for these bugs have an adverse impact on the web framework I'm developing, which relies heavily on autocompletion suggestions for helping the user correctly configure UI components.

Nevertheless, there is something important that I believe you should be aware of:
Although you've concluded that these are bugs in the typescript language service, I'm not so sure about that!
Would you please try the example above using the `typescript-atomizer` Atom package? You will see that **it correctly displays autocomplete suggestions for all the test cases I've mentioned**, and it does it either as soon as I type a dot, or when I press `ctrl+space`.
One possibility that could explain the different behavior is that `typescript-atomizer` is using an outdated compiler version (so it has no support for the latest 1.4 language features, and that's one of the reasons I'm not using it, by the way).
Perhaps that compiler version behaves differently and it doesn't have these bugs (i.e. these are regressions), or perhaps `typescript-atomizer` is doing something different that you could also implement in your package.
Either way, I believe it could be worth to investigate, would you agree?

**Comment by basarat** [comment]
> believe it could be worth to investigate, would you agree?

Good idea. Seem to use the same endpoint that we are using : https://github.com/timjroberts/typescript-atomizer-deploy/blob/506dc9675f2fff3d0a3d9adf56e7ebbdcb09b381/lib/TypeScriptTextEditor.js#L269 

Ours: https://github.com/TypeStrong/atom-typescript/blob/master/lib/main/lang/projectService.ts#L133

Might just be language service versions (highly likely to be honest). But might also be the extra work we do on the completions. Will investigate more.

**Comment by claudio-silva** [comment]
I can see `typescript-atomizer` is calling `getCompletionsAtPosition` with `true` for the last parameter (`isMemberCompletion`) while your code isn't. Could that have any influence on the returned list?

**Comment by basarat** [comment]
That's just a difference of language service versions (our being newer). Our's doesn't take the parameter

![image](https://cloud.githubusercontent.com/assets/874898/6205859/827485e4-b5d0-11e4-8e5a-e16edff3bebe.png)

**Comment by basarat** [comment]
Fixed with `0.55.0`. Thanks for the report + testing + patience. Let us know if anything else doesn't work :+1:

**Comment by claudio-silva** [comment]
Thank you very much for acting so quickly on this! :thumbsup:
This was a major blocker for me, so I'm quite glad to, finally, being able to start working with your package for my development work. :smile:
